egs/librispeech/ASR/long_file_recog/merge_chunks.py

```diff
@@ -0,0 +1,181 @@
+# limitations under the License.
+
+
+"""
+This file merge overlapped chunks into utterances accroding to recording ids.
+"""
+
+import argparse
+import logging
+from pathlib import Path
+from cytoolz.itertoolz import groupby
+
+import sentencepiece as spm
+from lhotse import CutSet, SupervisionSet, load_manifest
+from lhotse import SupervisionSegment, MonoCut
+
+
+def get_parser():
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument(
+        "--bpe-model",
+        type=str,
+        default="data/lang_bpe_500/bpe.model",
+        help="Path to the BPE model",
+    )
+
+    parser.add_argument(
+        "--manifest-in-dir",
+        type=Path,
+        default=Path("data/librilight/manifests_chunk_recog"),
+        help="Path to directory of chunk cuts with recognition results.",
+    )
+
+    parser.add_argument(
+        "--manifest-out-dir",
+        type=Path,
+        default=Path("data/manifests"),
+        help="Path to directory to save full utterance by merging overlapped chunks.",
+    )
+
+    parser.add_argument(
+        "--extra",
+        type=float,
+        default=2.0,
+        help="""Extra duration (in seconds) at both sides.""",
+    )
+
+    return parser.parse_args()
+
+
+def merge_chunks(
+    cuts_chunk: CutSet,
+    supervisions: SupervisionSet,
+    sp: spm.SentencePieceProcessor,
+    extra: float,
+) -> CutSet:
+    """Merge chunk-wise cuts accroding to recording ids.
+
+    Args:
+      cuts_chunk:
+        The chunk-wise cuts.
+      supervisions:
+        The supervision manifest containing text file path.
+      sp:
+        The BPE model.
+      extra:
+        Extra duration (in seconds) to drop at both sides of each chunk.
+    """
+    # Divide into groups accroding to their recording ids
+    cut_groups = groupby(lambda cut: cut.recording.id, cuts_chunk)
+
+    utt_cut_list = []
+    for recording_id, cuts in cut_groups.items():
+        # For each group with a same recording, sort it accroding to the start time
+        chunk_cuts = sorted(cuts, key=(lambda cut: cut.start))
+
+        rec = chunk_cuts[0].recording
+        alignments = []
+        cur_end = 0
+        for cut in chunk_cuts:
+            # Get left and right borders
+            left = cut.start + extra if cut.start > 0 else 0
+            chunk_end = cut.start + cut.duration
+            right = chunk_end - extra if chunk_end < rec.duration else rec.duration
+
+            # Assert the chunks are continuous
+            assert left == cur_end, (left, cur_end)
+            cur_end = right
+
+            assert len(cut.supervisions) == 1, len(cut.supervisions)
+            for ali in cut.supervisions[0].alignment["symbol"]:
+                t = ali.start + cut.start
+                if left <= t < right:
+                    alignments.append(ali.with_offset(cut.start))
+
+        old_sup = supervisions[rec.id]
+        assert old_sup.recording_id == rec.id, (old_sup.recording_id, rec.id)
```

Comment:
Given the token-level alignments, if I want to get corresponding word-level alignments (for example, to create a CTM file), is there a way to do that? At the moment, I get the full output text as:
```python
        # Decode the BPE tokens to text
        hyp = [ali.symbol for ali in alignments]
        hyp_text = sp.decode(hyp)
        hyp_text = asr_text_post_processing(hyp_text)
```
Basically, the problems comes down to knowing the mapping that `sp.decode` creates, but I don't know if this mapping is accessible.

Reply:
Tagging @csukuangfj since you may be familiar with sentencepiece.

Reply:
Found this relevant function: https://github.com/k2-fsa/icefall/blob/30bde4b788f0ad05110f6981b630b745ce6ffb82/egs/librispeech/ASR/transducer_stateless/alignment.py#L239

Reply:
You could use this https://github.com/k2-fsa/icefall/blob/30bde4b788f0ad05110f6981b630b745ce6ffb82/icefall/utils.py#L1392. It would cover more cases.

Reply:
@yaozengwei Thanks! This is exactly what I needed :)